mesonbuild/environment.py

```diff
@@ -403,12 +403,22 @@ def detect_cpu(compilers: CompilersDict) -> str:
                                         'linux': 'linux',
                                         'cygwin': 'nt',
                                         'darwin': 'xnu',
-                                        'sunos': 'sunos',
                                         'dragonfly': 'dragonfly',
                                         'haiku': 'haiku',
                                         }
 
 def detect_kernel(system: str) -> T.Optional[str]:
+    if system == 'sunos':
+        # This needs to be /usr/bin/uname because gnu-uname could be installed and
+        # won't provide the necessary information
+        p, out, _ = Popen_safe(['/usr/bin/uname', '-o'])
+        if p.returncode != 0:
+            raise MesonException('Failed to run "/usr/bin/uname -o"')
+        out = out.lower().strip()
+        if out not in {'illumos', 'solaris'}:
+            mlog.warning(f'Got an unexpected value for kernel on a SunOS derived platform, expcted either "illumos" or "solaris", but got "{out}".'
```

Comment:
Maybe we should just return None here instead of out?

Reply:
Yeah, that makes sense